### ISSUE TITLE
CClientScript ignores DynamicCaching

### ISSUE THREAD

**Issue by marcovtwout** [yiisoft/yii#2384]
``` php
...other HTML content...
<?php if($this->beginCache($id)) { ?>
...fragment content to be cached...
    <?php $this->renderDynamic($callback); ?>
...fragment content to be cached...
<?php $this->endCache(); } ?>
...other HTML content...
```

In any of the register*()-functions of CClientScript, a caching action is recorded.
This is (incorrectly) also happening when renderDynamic() is handling it's callback.

**Expected**
Any calls to CClientScript from renderDynamic($callback) should not be cached and replayed.

**Comment by degibons** [comment]
faced with this problem too, please fix it

**Comment by cebe** [comment]
> In any of the register*()-functions of CClientScript, a caching action is recorded.
> This is (incorrectly) also happening when renderDynamic() is handling it's callback.

can you be a bit more specific about this part? I do not really get the problem.

**Comment by marcovtwout** [comment]
It's been a while since I posted this, and I currently cannot reproduce it. @degibons could you provide an example?

**Comment by degibons** [comment]
Ok, my example:

controller:
------------------------------
```php
public function filters()
{
        return array(
            array(
                'COutputCache + index',
            ),
        );
}

public function actionIndex()
{
  $this->render('view');
}

public function myFunc()
{
    if ($someBoolean)
        Yii::app()->clientScript->registerScript('myScript1', $myScript1)
    else
        Yii::app()->clientScript->registerScript('myScript2', $myScript2)
}
```
------------------------------

view.twig:
-----------------------------
```twig
some cached html...
this.renderDynamic('myFunc')
some cached html...
```
------------------------------

So, in this example if at first run $someBoolean=true $myScript1 will be registered on the page. But if at second run $someBoolean=false $myScript1 will still be registered. It happens because all clientScript->register*** methods have recordCachingAction() call, no matter was it ordinary call or inside renderDynamic().

**Comment by PatchRanger** [comment]
I am facing quite different symptoms but I am sure they have the same roots as this issue.
My problem is that `dynamicContent` gives what it's intended to give only before its result is placed to cache. After that it gives always the same result - which quickly becomes incorrect - though the cache has the dynamic fragment escaped (`<###dynamic-$n###>`).

Why do I think our issues are related?
Because the root of the problem is the fact that `dynamicContent` is replaced to its value at the moment of placing to cache while it's intended to be replaced at each page render (that is why it is called `dynamicContent`, I guess).

I am going to provide a pull-request for that soon, stay tuned))

**Comment by PatchRanger** [comment]
@cebe How can I force merging of the related pull-request? Is the issue clear? And is the provided solution correct?
The situation is that I've pushed that commit to `master` branch of my fork (my fault) - so now I can't propose any pull-requests, I am stuck))

**Comment by cebe** [comment]
@PatchRanger you can create a new branch from your current master `git checkout -b newbranch` and send a new pull request from that branch. after you have done that you can reset your master.

**Comment by PatchRanger** [comment]
@cebe Thanks, it worked.

**Comment by degibons** [comment]
Does this bug will ever be verified?

**Comment by samdark** [comment]
We're kinda focused on Yii 2.0 nowadays. A failing unit test would be absolute confirmation...

**Comment by samdark** [comment]
Won't be fixed in 1.1.
